### ISSUE TITLE
⚠️ Using both asset pipeline and webpacker is not thread safe

### ISSUE THREAD

**Issue by kylefox** [jamesmartin/inline_svg#112]
Using both `inline_svg_tag` (asset pipeline) and `inline_svg_pack_tag` (webpacker) in the same application can cause big problems 😬

The helpers were sporadically unable to find the specified SVGs. Sometimes they would work fine, other times they would render the placeholder (`<svg><!-- SVG file not found: logo.svg --></svg>`).

After much investigation, we determined that [the changes that added `inline_svg_tag` and `inline_svg_pack_tag`](https://github.com/jamesmartin/inline_svg/pull/103) (made by me 🤭) **are not thread safe** — specifically, changing `InlineSvg.configuration.asset_finder` within [`with_asset_finder`](https://github.com/jamesmartin/inline_svg/pull/103/files#diff-76810304c6563797234460d5732ea6b3R71-R79) can create a race condition.

This was a severe problem for us, so I [hotfixed it](https://github.com/kylefox/inline_svg/commit/21a9e072f6ea88b044172cfe95dd90ee9065b47a) by using a thread-safe variable (`Thread.current[:inline_svg_asset_finder]`) to set/retrieve the asset finder instead of changing/using `InlineSvg.configuration.asset_finder` directly. It's admittedly a gross hack, but seems to have fixed our issue.

Before I submit this as a PR, I'm wondering if anyone knows of a better fix — cc @jamesmartin?

**Comment by kylefox** [comment]
Another idea I had was to modify some of the internals to accept a finder as an optional argument.

For example:

```ruby
module InlineSvg
  class FindsAssetPaths
    def self.by_filename(filename)
      asset = configured_asset_finder.find_asset(filename)
      asset.try(:pathname) || asset.try(:filename)
    end

    def self.configured_asset_finder
      InlineSvg.configuration.asset_finder
    end
  end
end
```

becomes

```ruby
module InlineSvg
  class FindsAssetPaths
    def self.by_filename(filename, asset_finder: InlineSvg.configuration.asset_finder)
      asset = asset_finder.find_asset(filename)
      asset.try(:pathname) || asset.try(:filename)
    end
  end
end
```

It would mean changing some internal API calls, but I think it could be done in a backwards-compatible way. It's more explicit and probably a cleaner approach than setting thread-specific variables.

**Comment by jamesmartin** [comment]
@kylefox thanks for the write-up of this problem and suggestions for fixes. ✨ 

I like the idea of passing the `asset_finder` around. This should allow the Webpack and sprockets implementations to diverge where necessary.

**Comment by kylefox** [comment]
> It would mean changing some internal API calls, but I think it could be done in a backwards-compatible way.

Hmmm ... after looking at the code more, I'm not sure it's possible to remain totally backwards-compatible. 🤔 In order to pass the `asset_finder` all the wait down the chain, the method signature for asset file classes would need to change from `named(asset_name)` to `named(asset_name, asset_finder)`.

This is fine for built-in asset files ([`InlineSvg::AssetFile.named`](https://github.com/jamesmartin/inline_svg/blob/master/lib/inline_svg/asset_file.rb#L6-L11) and [`InlineSvg::CachedAssetFile#named`](https://github.com/jamesmartin/inline_svg/blob/master/lib/inline_svg/cached_asset_file.rb#L28-L31)), but _does_ present a problem for [custom asset file loaders](https://github.com/jamesmartin/inline_svg#custom-asset-file-loader).

I'm wondering if we could:

1. Release a patch version using the aforementioned `Thread.current` fix. It's a bit ugly, but it works — and is arguably less ugly than sporadically disappearing SVGs 😕 
2. Consider a new API for asset files/loaders in [Version 2.0](https://github.com/jamesmartin/inline_svg/issues/1050) that eliminates the need for the `Thread.current` fix. Perhaps the internal and external APIs can be simplified, now that webpack is officially supported by Rails and this library supports both Sprockets and Webpacker simultaneously.

Thoughts @jamesmartin?

**Comment by jamesmartin** [comment]
> I'm wondering if we could:
> 
> 1. Release a patch version using the aforementioned `Thread.current` fix. It's a bit ugly, but it works — and is arguably less ugly than sporadically disappearing SVGs 😕
> 2. Consider a new API for asset files/loaders in [Version 2.0](https://github.com/jamesmartin/inline_svg/issues/1050) that eliminates the need for the `Thread.current` fix. Perhaps the internal and external APIs can be simplified, now that webpack is officially supported by Rails and this library supports both Sprockets and Webpacker simultaneously.

@kylefox this sounds like a sensible way forward. Do you want to open a PR with your `Thread.local` hotfix so I can get that out as a patch?

**Comment by kylefox** [comment]
Done: https://github.com/jamesmartin/inline_svg/pull/114

**Comment by jamesmartin** [comment]
Fix released in [v1.7.0](https://github.com/jamesmartin/inline_svg/releases/tag/v1.7.0).
